juju/juju
# 2.7 Upgrade for Systemd File Relocation

### Checklist


----

## Description of change

This patch adds a 2.7.2 upgrade step to relocate systemd unit files for agents.

The `UpgradableService` indirection has a new method `RemoveOldService`, which is implemented by `systemd.Service`. It removes the legacy service directory from _/lib/systemd/system_.

This method is called by the service manager when writing files, so it will work for all prior upgrade steps using the service manager to rewrite service units.

Refactoring includes adding a shim for file-system operations and removing some of the patching-mocks-over-function-vars messiness.

## QA steps

- Bootstrap with a prior Juju (2.6 will do).
- `juju deploy cs:~jameinel/ubuntu-lite`
- When settled, build this patch and run `juju upgrade-controller --build-agent`.
- `juju upgrade-model`.
- Check that everything is online.
- `juju ssh 0`. Check that _/etc/systemd/system_ contains service definitions and start scripts for the unit and machine, and there are no directories for them in _/lib/systemd/system_.

I also successfully tested a series upgrade from Trusty to Xenial, but this should not be required - it is covered by existing functional tests.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1860432
